### PR TITLE
Add Jib Ownership Maven Extension

### PR DESCRIPTION
**Pull request by chanseokoh**
Towards https://github.com/GoogleContainerTools/jib/pull/1257 and https://github.com/GoogleContainerTools/jib/pull/1955, as well as a workaround for https://github.com/GoogleContainerTools/jib/pull/1270.

The Jib Maven extension allows the user to customize file ownership in layers.

The following example sets UID 300 and GID 500 for `/app/classes/**`.
```xml
<configuration>
  <pluginExtensions>
    <pluginExtension>
      <implementation>com.google.cloud.tools.jib.plugins.extension.maven.ownership.JibOwnershipExtension</implementation>
      <configuration implementation="com.google.cloud.tools.jib.plugins.extension.maven.ownership.Configuration">
        <rules>
          <rule>
            <glob>/app/classes/**</glob>
            <ownership>300:500</ownership>
          </rule>
        </rules>
      </configuration>
    </pluginExtension>
```

However, due to [the way Jib adding parent directory tarball entries](https://github.com/GoogleContainerTools/jib/issues/1270), the directories `/app` and `/app/classes` will continue to have the ownership `0:0` (`root:root`). Considering the [use-case of Skaffold](https://github.com/GoogleContainerTools/jib/issues/1257#issuecomment-571586488), we should verify if this can cover the use-case; otherwise, we may need to revisit https://github.com/GoogleContainerTools/jib/issues/1270.

### CODE REVIEW

File: first-party/ownership-extension/maven/src/main/java/com/google/cloud/tools/jib/plugins/extension/maven/ownership/JibOwnershipExtension.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    List<FileEntry> entries =
+        layer.getEntries().stream().map(this::modifyFileEntry).collect(Collectors.toList());
+    return layer.toBuilder().setEntries(entries).build();
+  }
+
+  private FileEntry modifyFileEntry(FileEntry entry) {
+    for (Entry<PathMatcher, String> mapEntry : pathMatchers.entrySet()) {
+      PathMatcher matcher = mapEntry.getKey();
+      Path pathInContainer = Paths.get(entry.getExtractionPath().toString());
+      if (matcher.matches(pathInContainer)) {
```
Comment (chanseokoh) [inline first-party/ownership-extension/maven/src/main/java/com/google/cloud/tools/jib/plugins/extension/maven/ownership/JibOwnershipExtension.java:83]: note to self: the current code doesn't hurt, but I need only the last (or first?) `PathMatchers` that matches.

**Comment by loosebazooka** [comment]
can we change the PR description to describe the ownership extension?

File: first-party/ownership-extension/maven/src/main/java/com/google/cloud/tools/jib/plugins/extension/maven/ownership/Configuration.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.plugins.extension.maven.ownership;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (loosebazooka) [inline first-party/ownership-extension/maven/src/main/java/com/google/cloud/tools/jib/plugins/extension/maven/ownership/Configuration.java:22]: perhaps a description of config here?

File: first-party/ownership-extension/maven/src/main/java/com/google/cloud/tools/jib/plugins/extension/maven/ownership/JibOwnershipExtension.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    @SuppressWarnings("unchecked")
+    List<FileEntriesLayer> layers = (List<FileEntriesLayer>) buildPlan.getLayers();
+    List<FileEntriesLayer> newLayers =
+        layers.stream().map(this::modifyLayer).collect(Collectors.toList());
+    return buildPlan.toBuilder().setLayers(newLayers).build();
+  }
+
+  private FileEntriesLayer modifyLayer(FileEntriesLayer layer) {
```
Comment (loosebazooka) [inline first-party/ownership-extension/maven/src/main/java/com/google/cloud/tools/jib/plugins/extension/maven/ownership/JibOwnershipExtension.java:77]: I think modify technically isn't the process happening here? We're creating a new layer with changed values?

**Comment by chanseokoh** [comment]
Updated the description. Worth checking.

I'll take care of the code comments next week.

File: first-party/ownership-extension/maven/src/main/java/com/google/cloud/tools/jib/plugins/extension/maven/ownership/JibOwnershipExtension.java (hunk 3 above)
Comment (chanseokoh) [reply first-party/ownership-extension/maven/src/main/java/com/google/cloud/tools/jib/plugins/extension/maven/ownership/JibOwnershipExtension.java]: Changed the names to `applyRulesToLayer/FileEntry`. That said, I changed `Configuration.Entry` to `Configuration.Rule`. The latter sounds better.

**Pull request merged**